File: seminar-11/longest_increasing_subsequence/cpp/longest_increasing_subsequence.cpp
#include <vector>

using namespace std;

vector<int> longest_increasing_subsequence(const vector<int> &a) {
    int n = (int) a.size();
    vector<int> dp(n, 1);
    vector<int> prev(n, -1);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < i; j++) {
            if (a[j] < a[i] && dp[j] + 1 > dp[i]) {
                dp[i] = dp[j] + 1;
                prev[i] = j;
            }
        }
    }

    int pos = 0;
    for (int i = 1; i < n; i++) {
        if (dp[i] > dp[pos]) {
            pos = i;
        }
    }

    vector<int> ans;
    while (pos != -1) {
        ans.push_back(a[pos]);
        pos = prev[pos];
    }

    reverse(ans.begin(), ans.end());
    return ans;
}

File: seminar-11/longest_increasing_subsequence/python/longest_increasing_subsequence.py
def longest_increasing_subsequence(a):
    """
    a — список чисел
    возвращает одну из наибольших возрастающих подпоследовательностей
    """
    n = len(a)
    dp = [1] * n
    prev = [-1] * n

    for i in range(n):
        for j in range(i):
            if a[j] < a[i] and dp[j] + 1 > dp[i]:
                dp[i] = dp[j] + 1
                prev[i] = j

    pos = 0
    for i in range(1, n):
        if dp[i] > dp[pos]:
            pos = i

    ans = []
    while pos != -1:
        ans.append(a[pos])
        pos = prev[pos]

    ans.reverse()
    return ans
